Source/Urho3D/Input/DirectionalPadAdapter.cpp

```diff
@@ -0,0 +1,405 @@
+    if (axisIndex == 0)
+    {
+        if (value > 0.6f)
+        {
+            Append(right_, eventId, SCANCODE_RIGHT);
+            Remove(left_, eventId, SCANCODE_LEFT);
+        }
+        else if (value < -0.6f)
+        {
+            Append(left_, eventId, SCANCODE_LEFT);
+            Remove(right_, eventId, SCANCODE_RIGHT);
+        }
+        else if (value > -0.4f && value < 0.4f)
+        {
+            Remove(right_, eventId, SCANCODE_RIGHT);
+            Remove(left_, eventId, SCANCODE_LEFT);
+        }
+    }
+}
+
+void DirectionalPadAdapter::HandleJoystickHatMove(StringHash eventType, VariantMap& args)
+{
+    using namespace JoystickHatMove;
+    const auto hatIndex = args[P_HAT].GetUInt();
+    if (hatIndex != 0)
+        return;
+
+    const auto joystickId = args[P_JOYSTICKID].GetUInt();
+    const auto eventId = static_cast<InputType>(static_cast<unsigned>(InputType::JoystickDPad) + joystickId);
+
+    const auto position = args[P_POSITION].GetUInt();
+
+    if (0 != (position & HAT_UP))
+        Append(up_, eventId, SCANCODE_UP);
+    else
+        Remove(up_, eventId, SCANCODE_UP);
+
+    if (0 != (position & HAT_DOWN))
+        Append(down_, eventId, SCANCODE_DOWN);
+    else
+        Remove(down_, eventId, SCANCODE_DOWN);
+
+    if (0 != (position & HAT_LEFT))
+        Append(left_, eventId, SCANCODE_LEFT);
+    else
+        Remove(left_, eventId, SCANCODE_LEFT);
+
+    if (0 != (position & HAT_RIGHT))
+        Append(right_, eventId, SCANCODE_RIGHT);
+    else
+        Remove(right_, eventId, SCANCODE_RIGHT);
+
+    const float horizontalValue = ((position & HAT_RIGHT) ? 1.0f : 0.0f) + ((position & HAT_LEFT) ? -1.0f : 0.0f);
+    const float verticalValue = ((position & HAT_DOWN) ? 1.0f : 0.0f) + ((position & HAT_UP) ? -1.0f : 0.0f);
+}
+
+void DirectionalPadAdapter::HandleJoystickDisconnected(StringHash eventType, VariantMap& args)
+{
+    using namespace JoystickDisconnected;
+    auto joystickId = args[P_JOYSTICKID].GetUInt();
+
+    // Cancel Axis states.
+    const auto joyEventId = static_cast<InputType>(static_cast<unsigned>(InputType::JoystickAxis) + joystickId);
+    Remove(up_, joyEventId, SCANCODE_UP);
+    Remove(down_, joyEventId, SCANCODE_DOWN);
+    Remove(left_, joyEventId, SCANCODE_LEFT);
+    Remove(right_, joyEventId, SCANCODE_RIGHT);
+
+    // Cancel DPad states.
+    const auto dpadEventId = static_cast<InputType>(static_cast<unsigned>(InputType::JoystickDPad) + joystickId);
+    Remove(up_, dpadEventId, SCANCODE_UP);
+    Remove(down_, dpadEventId, SCANCODE_DOWN);
+    Remove(left_, dpadEventId, SCANCODE_LEFT);
+    Remove(right_, dpadEventId, SCANCODE_RIGHT);
+}
+
+void DirectionalPadAdapter::Append(InputVector& activeInput, InputType input, Scancode scancode)
+{
+    for (unsigned i = 0; i < activeInput.size(); ++i)
+    {
+        if (activeInput[i] == input)
+        {
+            return;
+        }
+    }
+    activeInput.push_back(input);
+    if (activeInput.size() == 1)
```

Comment:
Can you explain the behavior of adapter when multiple inputs are received?

Reply:
I should probably write a documentation on it.

DPadAdapter collects all inputs that it can categorize as a movement into specific direction. When at least one input received it sends a corresponding keyboard message about arrow key been pressed. When last input is released it sends a message about key been released.

So basically it's a message adapter that translates various input messages into discreet key event and state.